Show a warning when CRSes based on a datum ensemble are selected

nyalldawson opened the pull request:
Shows a warning in the projection selection widget when a CRS based on a datum ensemble is selected, warning the user that
there's an inherent lack of accuracy in the selected CRS

Requires PROJ 8+


https://user-images.githubusercontent.com/1829991/117229831-1efdae80-ae5f-11eb-8b18-8158aa97ebef.mp4

nyalldawson committed:
[needs-docs] Show a warning in the projection selection widget when
a CRS based on a datum ensemble is selected, warning the user that
there's an inherent lack of accuracy in the selected CRS

Requires PROJ 8+

nyalldawson commented:
@rouault @kbevers would love your eyes over this one!

kbevers commented:
I think the idea is good. You are very likely to confuse a lot of people with this so it would be nice to link to some doc page or external article that explains the deal with datum ensembles. 

I'm glad you included both WGS84 and ETRS89 in your example video. It sort of highlights that not all ensembles are bad. Sure, it is possible to achieve better accuracy than the 0.1 m of the ensemble with a specific ETRS89 realization but generally ETRS89 data agrees quite well with each other. Especially within the same country. The risk of doing something unwanted is not as high as with WGS84. The main issue for ETRS89 is that many users does not distinguish between WGS84 and ETRS89. With that in mind I am partial to only issueing this warning for WGS84 or perhaps when the ensemble accuracy exceeds 1.0 m (or whatever is best suited, some analysis is needed).

nyalldawson commented:
Thanks for the insight @kbevers !

> perhaps when the ensemble accuracy exceeds 1.0 m

Isn't that rather artificial? +/- 1.0 m in many disciplines is certainly worthy of a warning..! Even +/- 0.1m can be problematic. I'd rather not make any call here and leave it entirely in the user's hands to decide what level of inaccuracy is acceptable for their own work.

nyalldawson commented:
Here's a close up of the tooltip

![image](https://user-images.githubusercontent.com/1829991/117254984-75351680-ae8c-11eb-8844-560e3046ef61.png)

kbevers commented:
Just a clarification before I ramble on about datum ensembles in general :) When is this warning actually issued? I can think of four situations:

A. From a local CRS (let's say the Swedish SWEREF99) to a datum ensemble (WGS84)
B. From a datum ensemble (WGS84) to local CRS (SWEREF99)
C. From a datum ensemble (ETRS89) to a datum ensemble (WGS84)
D. From a datum ensemble (WGS84) to the same datum ensemble (WGS84 / UTM32)

A) is the worst scenario with a real risk of deterioating the data. B) is mostly problematic because the data will get a CRS label that implies better accuracy than is actually the case. C) Unintended thing can definately happen here. D) No problem at all, garbage in and garbage out (or maybe top notch data in and top notch data out, who knows).

I think warnings are in order for cases A, B and C. It may be a good idea to change the wording depending on the situation.

rouault commented:
Case D in above Kristian's comment is also a situation where no datum change occurs, so there's no inaccuracy beyond the least significant bit of a float64 occurring here. I kind of feel the warning would be legitimate each time there's a datum change. Or, more subtle, when there's a datum change where the transformation has a (in)accuracy > 0  (the transformation between French RGF93 and ETRS89, or SWEREF99 and ETRS89 are noop of accuracy 0 m)

kbevers commented:
> Or, more subtle, when there's a datum change where the transformation has a (in)accuracy > 0 (the transformation between French RGF93 and ETRS89, or SWEREF99 and ETRS89 are noop of accuracy 0 m)

I would rather warn when there's a null transform in the mix. That is usually when something unexpected is happening, e.g. transforming from ETRS89 to WGS84. In reality you don't get what you ask for. Luckily I think most of the null transforms are between datum ensembles and hence is already covered by Nyalls implementation in this PR.

rouault commented:
> That is usually when something unexpected is happening, e.g. transforming from ETRS89 to WGS84.

The transformation between French RGF93 and ETRS89, or SWEREF99 and ETRS89 are EPSG registered transformation with a 0 Helmert shift and a 0m inaccuracy
The transformation between ETRS89 and WGS84 is a EPSG registered transformation with a 0 Helmert shift and a 1m inaccuracy, so you'd get the warning (the warning would be for > 0m inaccuracy or unknown/ballpark)

Anyway, this was to brainstorm a bit. I don't have a strong opinion regarding this. The accuracy of the data would also be a criterion. I you work with a Natural Earth level 0 shapefile, you likely don't care about errors of a few meters.

wonder-sk commented:
This is a bold move :grin:  I agree we should try to educate users, but still this warning seems quite scary. For the last 15 years of "doing GIS" I have been happily ignoring the fact that WGS 84 is a datum ensemble with limited accuracy, and it was not a problem so far :slightly_smiling_face: 

My worries are that (for now discussing just WGS 84):
- it is really hard to find introductory material for people to learn basics of the various realizations and how to deal with them
- the support with EPSG database is quite poor (AFAIK) - missing codes for projections, e.g. "WGS 84 (G1762) / UTM zone 33N", and building custom WKTs for those is far from trivial. I guess using just "WGS 84 / UTM zone 33N" would trigger the warning?
- I have not seen (yet) any data that would actually use a particular WGS 84 realization as the CRS
- support in other software/hardware is poor. For example, one practical problem: how do I know realization of WGS 84 data captured with my GPS (especially if it blends data from various GNSS)
- well-known projects like OpenStreetMap seem to embrace WGS 84 without any particular realization (?)

I know all this may be a bit of a chicken & egg problem, just wondering if this is not going to create a lot of confusion. And we could argue that many people do not really care about +/- 2 meters accuracy (especially if they opened QGIS just to view a shapefile :slightly_smiling_face:  )

rouault commented:
>     * the support with EPSG database is quite poor (AFAIK) - missing codes for projections, e.g. "WGS 84 (G1762) / UTM zone 33N", and building custom WKTs for those is far from trivial

The issue is that even if there was a EPSG code for WGS 84 (G1762) / UTM zone 33N", you need to store the coordinate epoch. A position acquired today in Australia in WGS 84 (G1762) will be off by 2 meters in 30 years. So maybe we should have a warning each time we use a dynamic (meaning non-plate fixed) datum (such as WGS 84, ITRFxxxx, etc.) and we don't store the coordinate epoch (which will always be the case currently) ?

> well-known projects like OpenStreetMap seem to embrace WGS 84 without any particular realization (?)

They use WGS 84 as mostly meaning "geographic coordinates in some datum", where "some datum" is generally the national datum, most of them plate fixed, that is in effect when the coordinate is added to the database (like most coordinates in France will be referenced against RGF93 and not WGS 84)

nyalldawson commented:
I'll reply to the comments later today, but for now I've split off the api component into a separate PR - https://github.com/qgis/QGIS/pull/43109 . There should be no controversy with that component alone :rofl:

nyalldawson commented:
@wonder-sk 
> This is a bold move grin I agree we should try to educate users, but still this warning seems quite scary. For the last 15 years of "doing GIS" I have been happily ignoring the fact that WGS 84 is a datum ensemble with limited accuracy, and it was not a problem so far slightly_smiling_face

Well... that you know of, anyway! :rofl:  I've different experience (which I can't detail too much about), where this resulted in weeks of wasted work, and financial/reputational embarrassment to an organisation. I really think we should plan that QGIS is used for high-accuracy-requirement work, and then let those who work in lower-accuracy-requirement fields choose to ignore issues which don't affect their work.

> it is really hard to find introductory material for people to learn basics of the various realizations and how to deal with them

Completely agree. It's not a QGIS specific issue though (or even a uniquely OS geo issue!) But I would suggest we could ask osgeo to look into funding this kind of introductory documentation so that it's available for users across the whole ecosystem (ie QGIS users, R users, etc). The underlying problem affects us all.

> the support with EPSG database is quite poor (AFAIK) - missing codes for projections, e.g. "WGS 84 (G1762) / UTM zone 

True, but I don't see this as a blocker. There's good choices available depending on your region (e.g. in Australia we can/should use GDA2020 (or 94). And Australia definitely isn't alone here...

>  I guess using just "WGS 84 / UTM zone 33N" would trigger the warning?

Correct -- the underlying datum is still an ensemble

> I have not seen (yet) any data that would actually use a particular WGS 84 realization as the CRS

To pick one example that I know well: GDA94/2020

> support in other software/hardware is poor. For example, one practical problem: how do I know realization of WGS 84 data captured with my GPS (especially if it blends data from various GNSS)

Yep, definitely a big issue -- but again, I don't think we should ignore what we CAN do (inform and educate users) because of factors like this which we can't control

> well-known projects like OpenStreetMap seem to embrace WGS 84 without any particular realization (?)

See https://www.openstreetmap.org/user/StephaneP/diary/390290

> I know all this may be a bit of a chicken & egg problem, just wondering if this is not going to create a lot of confusion. And we could argue that many people do not really care about +/- 2 meters accuracy (especially if they opened QGIS just to view a shapefile slightly_smiling_face )

If they are just viewing then they'll never see the warning, but point taken. See my follow up responses below...

nyalldawson commented:
Ok, here's what I propose as a refinement:

- We make the show warning messages opt-in **API** for the widget, not opt-out (as it originally was)
- We show the warnings in these selected places, where the user is explicitly creating/modifying data:
   - Align raster dialog
   - DXF export dialog
   - Vector layer export dialog
   - Processing CRS widgets (possible we should introduce api to allow this to be controlled on an algorithm/parameter level), but initially I'd just show for all of them
   - Raster layer export dialog
   - DWG import dialog
   - New Geopackage/scratch/shapefile/spatialite layer dialog
   - Raster calculator
 - For raster layer/vector layer export dialogs, we don't show the warning when going from an original layer ensemble to the same destination ensemble. I.e. we show the warning if going from no ensemble -> ensemble, or when going from one ensemble -> another.
 - In the global settings, CRS tab, we introduce a setting "Don't show accuracy warnings for inaccuracies <= xxxx meters". Defaults to 0 so that new users see the warnings by default, but if they work in a field where +/- 20m is fine, then they can manually set this value to control when warnings are shown. (And then the consequences are firmly in their hands!). Organisations can control this setting in internal deployments via the global settings ini/startup scripts.

Thoughts?

rouault commented:
>     * we introduce a setting "Don't show accuracy warnings for inaccuracies <= xxxx meters". Defaults to 0

To avoid double negation, perhaps: "Show accuracy warnings for inaccuracies <= xxxx meters". Defaults to 1

nyalldawson commented:
> Show accuracy warnings for inaccuracies <= xxxx meters

Shouldn't that be >=?

rouault commented:
> Shouldn't that be >=?

of course. This demonstrates how hard it is to properly negate a sentence ;-)

nyalldawson commented:
Further refinement to the logic: when a geojson or kml or other format which mandates epsg4326 outputs is selected, the warning text will change to "The GeoJSON format requires EPSG:4326, which is based on a ensemble datum with accuracy of no better than ... Use an alternative format if better accuracy is required".

nyalldawson committed:
Make datum ensemble widget warnings opt-in, not opt-out

nyalldawson committed:
Show datum ensemble warnings for selected dialogs only

- align raster
- dxf export
- vector layer export
- processing crs widgets
- raster layer export
- dwg import
- new geopackage/shapefile/scratch layer dialog
- raster calculator

nyalldawson committed:
Also show warning in new table dialog

nyalldawson committed:
Rename API for flexibility

nyalldawson committed:
Fix typo

nyalldawson committed:
Don't show crs accuracy warnings when saving a vector/raster layer
from an ensemble based crs to another crs with the same ensemble datum

nyalldawson committed:
Add global setting to control the display of CRS inaccuracy warnings

This new setting in the Options -> CRS tab specifies a threshold
in meters, where CRS accuracy warnings will only be shown for
inaccuracies which exceed this distance.

This allows users who work in fields where lower accuracies are
acceptable to hide the CRS inaccuracy warnings when an ensemble datum
based CRS (such as 4326, 3857) is selected.

nyalldawson commented:
Ok, the new logic is implemented. The only bit I haven't done is the refined warning when saving to GeoJSON/KML/GPX, as it turns out that we don't currently force these formats to use EPSG:4326, and I'm reluctant to change that without wider discussion.

nyalldawson commented:
@rouault @kbevers @wonder-sk any objections to the latest version?

rouault commented on src/gui/qgsprojectionselectionwidget.h:
```diff
@@ -100,6 +101,46 @@ class GUI_EXPORT QgsProjectionSelectionWidget : public QWidget
+     * \see sourceEnsemble()
+     * \since QGIS 3.20
+     */
+    void setSourceEnsemble( const QString &ensemble );
```
My remark is perhaps not relevant as I'm not super familiar with this widget, but I'm wondering if that couldn't be something automatically done by the existing setCrs() method ?

nyalldawson replied:
Hmmmm, good question. I'd prefer to keep them separate as I think there's use cases when we might have a source layer in some crs yet want to show a different crs as default in the widget -- especially if we want the widget to default to 4326/3857 for some uses.

rouault commented on src/gui/qgsprojectionselectionwidget.h:
```diff
@@ -100,6 +101,46 @@ class GUI_EXPORT QgsProjectionSelectionWidget : public QWidget
+     * \see setSourceEnsemble()
+     * \since QGIS 3.20
+     */
+    QString sourceEnsemble() const;
```
does that need to be public, or exist at all ? It doesn't seem there are users of that method.

nyalldawson replied:
Not strictly, but I think it's good practice to have getters for all properties we can set. I've been hurt in the past when writing plugins targeting older qgis versions and there's missing getters -- it can be a real roadblock for plugin development.
